### PR TITLE
[WIP] transformed incremental + Rotations & Permutations [2]

### PR DESCRIPTION
**Pull request by TLESORT**

### CODE REVIEW

File: continuum/scenarios/transformation_incremental.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    ):
+        super(TransformationIncremental, self).__init__(
+            cl_dataset=cl_dataset,
+            nb_tasks=nb_tasks,
+            base_transformations=base_transformations
+        )
+        if incremental_transformations is None:
+            raise ValueError("For this scenario a list transformation should be set")
+
+        if nb_tasks != len(incremental_transformations):
```
Comment (arthurdouillard) [inline continuum/scenarios/transformation_incremental.py:40]: We should remove the argument `nb_tasks` here because it is strictly equivalent in all cases to the number of transformations. So it can be inferred.

File: continuum/scenarios/transformation_incremental.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            raise ValueError("The number of tasks is not equal to the number of transformation")
+
+        self.inc_trsf = incremental_transformations
+        self._nb_tasks = self._setup(nb_tasks)
+
+    def get_task_index(self, nb_tasks, y):
+        # all tasks have all labels, only the transformation change
```
Comment (arthurdouillard) [inline continuum/scenarios/transformation_incremental.py:47]: Can we have an option to use either 'shared label space' (the case here) and the opposite case?

This paper https://arxiv.org/abs/2006.14769 has well defined this distinction in its section 2. and table 1.

File: continuum/scenarios/transformation_incremental.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        return y
+
+    def get_task_transformation(self, task_index):
+        return transforms.Compose(self.inc_trsf[task_index] + self.trsf.transforms)
+
+    def update_task_indexes(self, task_index):
+        new_t_ = np.ones(len(self.dataset[1])) * task_index
+        self.dataset = (self.dataset[0], self.dataset[1], new_t_)
+
+    def __getitem__(self, task_index):
```
Comment (arthurdouillard) [inline continuum/scenarios/transformation_incremental.py:57]: The other scenarios and the abstract class accept either a task index (int) or several task index (slice). The latter case is not supported here (and I think it's correct), but you should have a check if/except if `task_index` is a `slice` to avoid any misunderstanding.

File: continuum/scenarios/transformation_incremental.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+    def get_task_index(self, nb_tasks, y):
+        # all tasks have all labels, only the transformation change
+        return y
+
+    def get_task_transformation(self, task_index):
+        return transforms.Compose(self.inc_trsf[task_index] + self.trsf.transforms)
+
+    def update_task_indexes(self, task_index):
+        new_t_ = np.ones(len(self.dataset[1])) * task_index
```
Comment (arthurdouillard) [inline continuum/scenarios/transformation_incremental.py:54]: No need of the trailing underscore.

File: continuum/scenarios/transformation_incremental.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+        :param task_index: The unique index of a task, between 0 and len(loader) - 1.
+        :return: A train PyTorch's Datasets.
+        """
+        self.update_task_indexes(task_index)
+        train = self._select_data_by_task(task_index)
+        trsf = self.get_task_transformation(task_index)
+        train_dataset = TaskSet(*train, trsf, data_type=self.cl_dataset.data_type)
+
+        return train_dataset
```
Comment (arthurdouillard) [inline continuum/scenarios/transformation_incremental.py:68]: Why "train"? It could be well the test set also, no?

File: continuum/scenarios/rotations.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        for tuple_ in list_degrees:
+            if isinstance(tuple_, int):
+                min = tuple_
+                max = tuple_
+            elif len(tuple_) == 2:
+                min, max = tuple_
+            else:
+                # list_degrees should contains list of rotations with:
+                # only one angle
+                # or with (min,max)
```
Comment (arthurdouillard) [inline continuum/scenarios/rotations.py:43]: Put this comment on the previous line.

File: continuum/scenarios/rotations.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                raise ValueError("list_degrees is wrong")
```
Comment (arthurdouillard) [inline continuum/scenarios/rotations.py:44]: The comment message (a few lines before) is actually quite informative, you should use it in the exception message so that the user can understand his error w/o having to look at our code.

Furthermore add in the message the `list_degrees` variable content.

File: continuum/scenarios/rotations.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    def __init__(
+            self,
+            cl_dataset: _ContinuumDataset,
+            nb_tasks: int,
+            list_degrees: Union[List[Tuple], List[int]],
+            base_transformations: List[Callable] = None
+    ):
+
+        if len(list_degrees) != nb_tasks:
+            raise ValueError("The nb of tasks != with number of angles tuples set in the list")
```
Comment (arthurdouillard) [inline continuum/scenarios/rotations.py:30]: Give the variable contents, likewise:

`f"The nb of tasks ({nb_tasks}) != with number of angles tuples ({len(list_degrees)}) set in the list"`

Anything that can speed up the debugging for the user is a good thing.

File: continuum/scenarios/rotations.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+
+        list_transformations = []
```
Comment (arthurdouillard) [inline continuum/scenarios/rotations.py:32]: I'd prefer this code to be in a different method, and called from the `__init__`.

File: continuum/scenarios/permutations.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    :param base_transformations: List of transformation to apply to all tasks.
+    :param seed: initialization seed for the permutations.
+    """
+
+    def __init__(
+            self,
+            cl_dataset: _ContinuumDataset,
+            nb_tasks: int,
+            base_transformations: List[Callable] = None,
+            seed=0
```
Comment (arthurdouillard) [inline continuum/scenarios/permutations.py:28]: Let the users define all tasks permutations seeds, thus a `list` instead of an `int` with an argument `seeds` (no need to add prefix "list" as in the [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation)).

And it would mean we don't need `nb_tasks`.

And in this case `nb_tasks` would be equal to `len(seeds) + 1` because the first task has no permutations. But I'm willing to discuss about this last point.

File: continuum/scenarios/permutations.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    ):
+        list_transformations = []
+        g_cpu = torch.Generator()
+        list_seed = torch.randperm(1000, generator=g_cpu)[:nb_tasks]
```
Comment (arthurdouillard) [inline continuum/scenarios/permutations.py:32]: This code is not reproducible, because the `list_seed` is different every time. This variable should be given by the user.

File: continuum/scenarios/permutations.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    # We inverse permutation is after self.trsf because it is done an torch tensor
+    def get_task_transformation(self, task_index):
+        return transforms.Compose(self.trsf.transforms + self.inc_trsf[task_index])
+
+
+class PermutationTransform:
+    """Permutation transformers.
+
+    This transformer is initialized with a seed such as same seed = same permutation.
+    Seed 0 means no permutations
```
Comment (arthurdouillard) [inline continuum/scenarios/permutations.py:55]: Don't add those special cases like this.

File: continuum/scenarios/permutations.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+            base_transformations: List[Callable] = None,
+            seed=0
+    ):
+        list_transformations = []
+        g_cpu = torch.Generator()
+        list_seed = torch.randperm(1000, generator=g_cpu)[:nb_tasks]
+        self.seed = seed
+
+        # first task is not permuted
+        list_seed[0] = 0
```
Comment (arthurdouillard) [inline continuum/scenarios/permutations.py:36]: You could create an identy permutation for the first task.

File: continuum/task_set.py [hunk 14]
```diff
@@ -92,14 +126,21 @@ def get_sample(self, index: int) -> np.ndarray:
...
-        t = self.t[index]
+        y = self._y[index]
+        t = self._t[index]
 
         if self.trsf is not None:
             img = self.trsf(img)
 
         return img, y, t
 
+    def get_image(self, index):
```
Comment (arthurdouillard) [inline continuum/task_set.py:137]: Do we need this method? We can simply do `task_set[index][0]` for the same result.

File: continuum/task_set.py [hunk 15]
```diff
@@ -45,40 +49,70 @@ def add_memory(
...
+        for i, ind in enumerate(indexes):
+            # we need to use get item to have the transform used
+            img, y, t = self.__getitem__(ind)
+
+            if i == 0:
+                if len(list(img.shape)) == 2:
+                    size_image = [1] + list(img.shape)
+                else:
+                    size_image = list(img.shape)
+                batch = torch.zeros(([len(indexes)] + size_image))
```
Comment (arthurdouillard) [inline continuum/task_set.py:99]: If I recall right, it's very slow to grow a numpy or torch tensor likewise. Better aggregate them in a python list and call `torch.cat(THE_LIST)` as the end. (or `vstack` or `tensor` I forgot).

Plus this method will be way shorter likewise.

File: continuum/task_set.py [hunk 16]
```diff
@@ -45,40 +49,70 @@ def add_memory(
...
-        return self.x.shape[0]
+        return self._x.shape[0]
+
+    def rand_samples(self, nb_samples):
+
+        nb_tot_samples = self._x.shape[0]
+        indexes = np.random.randint(0, nb_tot_samples, nb_samples)
+        return self.get_samples_from_ind(indexes)
+
+    def get_samples_from_ind(self, indexes):
```
Comment (arthurdouillard) [inline continuum/task_set.py:85]: You could write this function in the __getitem__ and allows `index` to be (in addition of an int) a list or a slice. This would remove an additional method while staying pythonic.

File: continuum/task_set.py [hunk 17]
```diff
@@ -45,40 +49,70 @@ def add_memory(
...
         """
-        plot(self, title=title, path=path, nb_per_class=nb_per_class, shape=shape)
+        plot_samples(self, title=title, path=path, nb_samples=nb_samples, shape=shape)
 
     def __len__(self) -> int:
         """The amount of images in the current task."""
-        return self.x.shape[0]
+        return self._x.shape[0]
+
+    def rand_samples(self, nb_samples):
```
Comment (arthurdouillard) [inline continuum/task_set.py:79]: Function name should begin with a verb.

File: continuum/task_set.py [hunk 18]
```diff
@@ -109,19 +150,16 @@ def split_train_val(dataset: TaskSet, val_split: float = 0.1) -> Tuple[TaskSet,
...
     val_indexes = indexes[:int(val_split * len(indexes))]
 
-    x, y, t = dataset.x, dataset.y, dataset.t
-    train_dataset = TaskSet(
-        x[train_indexes], y[train_indexes], t[train_indexes], dataset.trsf, dataset.data_type
-    )
-    val_dataset = TaskSet(
-        x[val_indexes], y[val_indexes], t[val_indexes], dataset.trsf, dataset.data_type
-    )
+    x_train, y_train, t_train = dataset.get_raw_samples_from_ind(train_indexes)
```
Comment (arthurdouillard) [inline continuum/task_set.py:159]: Why not using directly `x` `y` `z`?

As long as we are in internal code, no need to wrap getter.

File: continuum/viz.py [hunk 19]
```diff
@@ -1,72 +1,101 @@
...
-    plt.title(title)
-
-    if path is None:
-        plt.show()
+
+def plot_samples(dataset, title="", path=None, nb_samples=100, shape=None):
+    batch, _, _ = dataset.rand_samples(nb_samples)
+    filename = os.path.join(path, title)
+
+    if shape is None:
```
Comment (arthurdouillard) [inline continuum/viz.py:11]: Cool trick that I like (but you free to not use it):

`shape = shape or batch[0].shape`.

File: continuum/viz.py [hunk 20]
```diff
@@ -1,72 +1,101 @@
...
-
-    plt.imshow(big_image, cmap=cmap)
-    plt.xticks([])
-    plt.yticks([])
-    plt.title(title)
-
-    if path is None:
-        plt.show()
+
+def plot_samples(dataset, title="", path=None, nb_samples=100, shape=None):
```
Comment (arthurdouillard) [inline continuum/viz.py:7]: I feel the code of this module is more complicated than before for few new features, was that necessary? (I'm not talking about the rand_samples stuff which is nice!)

File: tests/test_classorder.py [hunk 21]
```diff
@@ -9,7 +9,7 @@
 class InMemoryDatasetTest(InMemoryDataset):
 
     def __init__(self, *args, class_order=None, **kwargs):
-        super().__init__(*args, **kwargs)
+        super(InMemoryDatasetTest, self).__init__(*args, **kwargs)
```
Comment (arthurdouillard) [inline tests/test_classorder.py:12]: Why not `super()`?

File: continuum/task_set.py [hunk 22]
```diff
@@ -19,24 +20,27 @@ class TaskSet(TorchDataset):
...
+            self,
+            x: np.ndarray,
+            y: np.ndarray,
+            t: np.ndarray,
+            trsf: transforms.Compose,
+            data_type: str = "image_array"
     ):
-        self.x, self.y, self.t = x, y, t
+
+        self._x, self._y, self._t = x, y, t
```
Comment (arthurdouillard) [inline continuum/task_set.py:31]: What was the rational behind prefixing by `_` here?

File: tests/test_permutations.py [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+
+    return (x_train, y_train.astype(int)), (x_test, y_test.astype(int))
+
+'''
+Test the initialization with three tasks
+'''
+def test_init(numpy_data):
+    train, test = numpy_data
+    dummy = InMemoryDatasetTest(*train, *test)
+    clloader = Permutations(cl_dataset=dummy, nb_tasks=3, seed=0)
```
Comment (arthurdouillard) [inline tests/test_permutations.py:32]: You should check with a batch of size 1:
- that the image is actually changed between each task
- that by re-creating the dataset and clloader, we obtain the exact same permutations given the same seeds.

File: tests/test_rotations.py [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+    Trsf_1 = (15,20)
+    Trsf_2 = 45
+
+    list_degrees = [Trsf_0, Trsf_1, Trsf_2]
+
+    clloader = Rotations(cl_dataset=dummy, nb_tasks=3, list_degrees=list_degrees)
+
+
+    for task_id, train_dataset in enumerate(clloader):
+        continue
```
Comment (arthurdouillard) [inline tests/test_rotations.py:43]: Check if the rotation are actually made. You could create 2x2 images with a different digits per pixel and check one corner value at each task.

File: tests/test_transformation_incremental.py [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+    x, y = numpy_data
+    dummy = InMemoryDataset(x, y, train='train')
+
+    Trsf_0 = []
+    Trsf_1 = [transforms.RandomAffine(degrees=[40, 50])]
+    Trsf_2 = [transforms.RandomAffine(degrees=[85, 95])]
+
+    list_transf = [Trsf_0, Trsf_1, Trsf_2]
+
+    continuum = TransformationIncremental(cl_dataset=dummy, nb_tasks=3, incremental_transformations=list_transf)
```
Comment (arthurdouillard) [inline tests/test_transformation_incremental.py:36]: Test if affine is actually applied.

File: tests/test_visualization.py [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+from continuum.scenarios import Rotations
+from continuum.scenarios import Permutations
+from continuum.scenarios import ClassIncremental
+
+
+'''
+Test the visualization with three tasks
+'''
+def test_visualization_rotations():
+    clloader = Rotations(cl_dataset=MNIST("./pytest/Samples/Datasets", download=True, train=True),
```
Comment (arthurdouillard) [inline tests/test_visualization.py:15]: Can we do it with a dummy dataset? Downloading MNIST slows down the test pipeline.

File: continuum/scenarios/base.py [hunk 27]
```diff
@@ -88,13 +78,8 @@ def __getitem__(self, task_index: Union[int, slice]):
...
         :return: A train PyTorch's Datasets.
         """
-        data = self._select_data_by_task(task_index)
-
-        return TaskSet(
-            *data,
-            self.train_trsf if self.train else self.test_trsf,
-            data_type=self.cl_dataset.data_type
-        )
+        _x, _y, _t = self._select_data_by_task(task_index)
```
[the remark this answers is not in the document]
Comment (arthurdouillard) [reply continuum/scenarios/base.py:81]: I cannot put this comment at the right line, but look at this file lines 112-115. You should be able to remove them with your new PR.

File: continuum/datasets/base.py [hunk 28]
```diff
@@ -57,53 +57,44 @@ class PyTorchDataset(_ContinuumDataset):
...
-        dataset = self.dataset_type(self.data_path, download=False, train=train)
-        x, y = np.array(dataset.data), np.array(dataset.targets)
+        self.dataset = self.dataset_type(self.data_path, download=self.download, train=self.train)
 
+    def torch_dataset(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+        x, y = np.array(self.dataset.data), np.array(self.dataset.targets)
         return x, y, None
 
+    def get_data(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+        return self.torch_dataset()
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/datasets/base.py:72]: ok but then I can remove the function torch_dataset?

File: continuum/task_set.py (hunk 22 above)
Comment (TLESORT) [reply continuum/task_set.py:31]: I think we discussed it before, it is because they are private attribute

File: tests/test_rotations.py (hunk 24 above)
Comment (TLESORT) [reply tests/test_rotations.py:43]: It is checked in the visualization test, but  we can check it automatically

File: tests/test_visualization.py (hunk 26 above)
Comment (TLESORT) [reply tests/test_visualization.py:15]: hum, yes

File: continuum/datasets/base.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (arthurdouillard) [reply continuum/datasets/base.py:72]: Sure 👍

File: continuum/datasets/base.py [hunk 29]
```diff
@@ -122,24 +113,26 @@ class ImageFolderDataset(_ContinuumDataset):
...
     @property
     def data_type(self) -> str:
         return "image_path"
 
-    def init(self, train: bool) -> Tuple[np.ndarray, np.ndarray, Union[None, np.ndarray]]:
-        if train:
-            folder = self.train_folder
-        else:
-            folder = self.test_folder
+    def _download(self):
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/datasets/base.py:131]: fixed in 2cbb3c6

File: tests/test_visualization.py (hunk 26 above)
Comment (TLESORT) [reply tests/test_visualization.py:15]: I it is okay for you I will create a separate set of tests for the different datasets to evaluate if they works. We need to now, if there are bugs in download process or anything. we will not know it with dummy datasets only

File: continuum/datasets/base.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/datasets/base.py:72]: fixed in 7711df0

File: tests/test_taskset.py [hunk 30]
```diff
@@ -22,3 +22,13 @@ def test_split_train_val(val_split, nb_val):
     train_set, val_set = split_train_val(base_set, val_split)
     assert len(val_set) == nb_val
     assert len(train_set) + len(val_set) == len(base_set)
+
+
+def test_rand_samples():
```
Comment (TLESORT) [inline tests/test_taskset.py:27]: this test fail, it should be fixed

File: continuum/task_set.py (hunk 22 above)
Comment (TLESORT) [reply continuum/task_set.py:31]: can I mark this conv as "resolved"?

File: continuum/datasets/base.py [hunk 31]
```diff
@@ -122,24 +113,26 @@ class ImageFolderDataset(_ContinuumDataset):
...
-        super().__init__(download=download, **kwargs)
+    def __init__(self, data_folder: str, train: bool, download: bool = True, **kwargs):
+        super(ImageFolderDataset, self).__init__(download=download, **kwargs)
 
-        self.train_folder = train_folder
-        self.test_folder = test_folder
+        self.data_folder = data_folder
+        self.train = train
+
+        if download:
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/datasets/base.py]: fixed in ae877eb

File: continuum/datasets/base.py [hunk 32]
```diff
@@ -57,53 +57,44 @@ class PyTorchDataset(_ContinuumDataset):
...
+            t_: Union[None, np.ndarray] = None,
+            **kwargs
     ):
-        super().__init__(**kwargs)
-
-        self.train = (x_train, y_train, t_train)
-        self.test = (x_test, y_test, t_test)
+        self.data = (x_, y_, t_)
         self._data_type = data_type
+        super(InMemoryDataset, self).__init__(**kwargs)
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/datasets/base.py]: yes, I think it is not usefull here.
I fixed it in 7447cf3

File: continuum/datasets/base.py [hunk 33]
```diff
@@ -57,53 +57,44 @@ class PyTorchDataset(_ContinuumDataset):
     """
 
     # TODO: some datasets have a different structure, like SVHN for ex. Handle it.
-    def __init__(self, *args, dataset_type, **kwargs):
+    def __init__(self, *args, dataset_type, train: bool, **kwargs):
+        super(PyTorchDataset, self).__init__(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/datasets/base.py]: I am not sure but is the modif in 0f56574 what you meant ?

File: tests/test_visualization.py (hunk 26 above)
Comment (TLESORT) [reply tests/test_visualization.py:15]: If you agree, I will do it in another PR after finishing this one :)

File: tests/test_transformation_incremental.py (hunk 25 above)
Comment (TLESORT) [reply tests/test_transformation_incremental.py:69]: I started it in test_init but the test fail at the moment because of random_samples function in task_set class

File: continuum/task_set.py (hunk 22 above)
Comment (TLESORT) [reply continuum/task_set.py:31]: I will assume yes :)

File: tests/test_visualization.py (hunk 26 above)
Comment (TLESORT) [reply tests/test_visualization.py:16]: Finalement, je pense que pour les tests de visualisation c'est pas mal de regarder avec des vrais données. Ca donne une meilleur apperçu de si le pipeline fonctionne normalement. Du coup j'ai garder les test tel qu'ils etaient mais je les ai marqué comme "slow" et j'ai modifié le Makefile pour que travis tourne pas ces tests la. Faudra juste les tourner une fois de temps en temps pour vérifier que ça marche toujours

File: tests/test_permutations.py (hunk 23 above)
Comment (TLESORT) [reply tests/test_permutations.py]: Done

File: tests/test_transformation_incremental.py (hunk 25 above)
Comment (TLESORT) [reply tests/test_transformation_incremental.py:78]: done

File: tests/test_rotations.py (hunk 24 above)
Comment (TLESORT) [reply tests/test_rotations.py:43]: Done

File: continuum/datasets/fellowship.py [hunk 34]
```diff
@@ -14,35 +14,38 @@ def __init__(
         data_path: str = "",
         download: bool = True,
     ):
-        super().__init__(data_path, download)
+        super(Fellowship, self).__init__(data_path, download)
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/datasets/fellowship.py:17]: changed to super()

File: continuum/datasets/fellowship.py [hunk 35]
```diff
@@ -14,35 +14,38 @@ def __init__(
...
 
-        return x, y, None
+        return x, y, t
 
 
 class MNISTFellowship(Fellowship):
 
     def __init__(self, data_path: str = "", download: bool = True) -> None:
-        super().__init__([MNIST, FashionMNIST, KMNIST], data_path, download)
+        super(MNISTFellowship, self).__init__([MNIST, FashionMNIST, KMNIST], data_path, download)
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/datasets/fellowship.py:45]: changed to super()

File: continuum/datasets/imagenet.py [hunk 36]
```diff
@@ -38,48 +38,47 @@ class ImageNet100(ImageNet1000):
...
+            *args,
+            data_subset: Union[Tuple[np.array, np.array], str, None] = None,
+            **kwargs
     ):
-        self.train_subset = train_subset
-        self.test_subset = test_subset
-
-        super().__init__(*args, **kwargs)
+        self.data_subset = data_subset
+        super(ImageNet100, self).__init__(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/datasets/imagenet.py:47]: changed to super()

File: continuum/datasets/imagenet.py [hunk 37]
```diff
@@ -38,48 +38,47 @@ class ImageNet100(ImageNet1000):
...
 
     def _download(self):
-        super()._download()
+        super(ImageNet100, self)._download()
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/datasets/imagenet.py:50]: changed to super() in patch_pr29 branch

File: continuum/scenarios/class_incremental.py [hunk 38]
```diff
@@ -49,7 +45,8 @@ def __init__(
         self._nb_tasks = self._setup(nb_tasks)
 
     def _setup(self, nb_tasks: int) -> int:
-        x, y, _ = self.cl_dataset.init(train=self.train)
+
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/scenarios/class_incremental.py:48]: I have done auto-reformat I guess it solved it

File: continuum/scenarios/transformation_incremental.py (hunk 1 above)
Comment (TLESORT) [reply continuum/scenarios/transformation_incremental.py:40]: I think it helps to avoid some errors to have it a bit redundant here. I would keep the nb_tasks parameter

File: continuum/scenarios/transformation_incremental.py (hunk 2 above)
Comment (TLESORT) [reply continuum/scenarios/transformation_incremental.py:47]: 'shared label space' means that all tasks use the same labels ?

File: continuum/scenarios/transformation_incremental.py (hunk 5 above)
Comment (TLESORT) [reply continuum/scenarios/transformation_incremental.py:68]: yep, fixed in 2e45b12f9467cd84c64aceb10bc86786a62bfc56

File: continuum/scenarios/transformation_incremental.py (hunk 4 above)
Comment (TLESORT) [reply continuum/scenarios/transformation_incremental.py:54]: 2e45b12f9467cd84c64aceb10bc86786a62bfc56

File: continuum/datasets/fellowship.py [hunk 39]
```diff
@@ -14,35 +14,38 @@ def __init__(
...
 
 
 class CIFARFellowship(Fellowship):
 
     def __init__(self, data_path: str = "", download: bool = True) -> None:
-        super().__init__([CIFAR10, CIFAR100], data_path, download)
+        super(CIFARFellowship, self).__init__([CIFAR10, CIFAR100], data_path, download)
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/datasets/fellowship.py:51]: fixed

File: README.md [hunk 40]
```diff
@@ -23,11 +23,9 @@ from continuum import ClassIncremental, split_train_val
 from continuum.datasets import MNIST
 
 clloader = ClassIncremental(
-    MNIST("my/data/path", download=True),
+    MNIST("my/data/path", download=True, train=True),
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply README.md:26]: This is how it is done in pytorch dataset I think and the cl_loader should not care about test/train difference.

File: continuum/datasets/imagenet.py [hunk 41]
```diff
@@ -38,48 +38,47 @@ class ImageNet100(ImageNet1000):
...
-            self.train_subset = os.path.join(self.train_folder, "train_100.txt")
-            download(self.train_subset_url, self.train_folder)
-        if self.test_subset is None:
-            self.test_subset = os.path.join(self.test_folder, "val_100.txt")
-            download(self.test_subset_url, self.test_folder)
+        if self.data_subset is None:
+            self.data_subset = os.path.join(self.data_folder, filename)
+            download(self.subset_url, self.data_folder)
 
     def init(self, train: bool) -> Tuple[np.ndarray, np.ndarray, Union[np.ndarray, None]]:
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/datasets/imagenet.py:62]: Maybe we should create a issue about this, this problem should have appear in the tests

File: continuum/datasets/base.py [hunk 42]
```diff
@@ -8,15 +8,15 @@
...
         self.data_path = data_path
         self.download = download
+        self.train = train
 
         if self.download:
             self._download()
 
-    @abc.abstractmethod
-    def init(self, train: bool) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
+    def get_data(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/datasets/base.py:19]: cl scenarios use the get_data function. It is not an internal function in my mind

File: continuum/scenarios/base.py [hunk 43]
```diff
@@ -23,28 +23,18 @@ class _BaseCLLoader(abc.ABC):
...
-        self,
-        cl_dataset: _ContinuumDataset,
-        nb_tasks: int,
-        train_transformations: List[Callable] = None,
-        common_transformations: List[Callable] = None,
-        train=True
+            self,
+            cl_dataset: _ContinuumDataset,
+            nb_tasks: int,
+            base_transformations: List[Callable] = None
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/scenarios/base.py:29]: a don't have a strong opinion about it, but the "base_" prefix might help the goal of this argument

File: continuum/scenarios/base.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/scenarios/base.py:81]: 0dad97c

File: continuum/scenarios/class_incremental.py [hunk 44]
```diff
@@ -10,36 +10,32 @@
...
     :param nb_tasks: The scenario number of tasks
     :param increment: Either number of classes per task, or a list specifying for
                       every task the amount of new classes.
     :param initial_increment: A different task size applied only for the first task.
                               Desactivated if `increment` is a list.
-    :param train_transformations: A list of data augmentation applied to the train set.
-    :param common_transformations: A list of transformations applied to both the
-                                   the train set and the test set. i.e. normalization,
-                                   resizing, etc.
+    :param base_transformations: A list of transformations applied to all tasks
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/scenarios/class_incremental.py:21]: same as https://github.com/Continvvm/continuum/pull/29#discussion_r470547549

File: continuum/scenarios/data_incremental.py [hunk 45]
```diff
@@ -3,6 +3,7 @@
 
 class DataIncremental(_BaseCLLoader):
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/scenarios/data_incremental.py:4]: data incremental is NIC so mix between NI and NC. It is different, this loader is not much used in CL papers but we can add it in a later pr

File: continuum/scenarios/instance_incremental.py [hunk 46]
```diff
@@ -15,37 +15,28 @@ class InstanceIncremental(_BaseCLLoader):
     :param cl_dataset: A continual dataset.
     :param nb_tasks: The desired number of tasks. If left to 0, it will try to use
                      the dataset's default number of tasks.
-    :param train_transformations: A list of data augmentation applied to the train set.
-    :param common_transformations: A list of transformations applied to both the
-                                   the train set and the test set. i.e. normalization,
-                                   resizing, etc.
+    :param base_transformations: List of transformation to apply to all tasks.
```
[the remark this answers is not in the document]
Comment (TLESORT) [reply continuum/scenarios/instance_incremental.py:18]: same as https://github.com/Continvvm/continuum/pull/29#discussion_r470547549

File: continuum/scenarios/transformation_incremental.py (hunk 3 above)
Comment (TLESORT) [reply continuum/scenarios/transformation_incremental.py:57]: Can you do it and push on patch_pr29 branch ?

File: tests/test_classorder.py (hunk 21 above)
Comment (TLESORT) [reply tests/test_classorder.py:12]: replaced by super()

File: continuum/viz.py (hunk 20 above)
Comment (TLESORT) [reply continuum/viz.py:7]: maybe... I use this piece of code in other project and somehow it works. I had bugs with the previous version, it was an easy fix by copy/past.

File: continuum/scenarios/transformation_incremental.py (hunk 3 above)
Comment (arthurdouillard) [reply continuum/scenarios/transformation_incremental.py:57]: Done.

File: continuum/viz.py (hunk 19 above)
Comment (TLESORT) [reply continuum/viz.py:11]: I prefer the version with "if", easier to read (at least for me :) )

File: continuum/task_set.py (hunk 18 above)
Comment (TLESORT) [reply continuum/task_set.py:163]: Is this comment it still up to date?
In the current version, I used special name to differentiate train and val data

File: continuum/task_set.py (hunk 17 above)
Comment (TLESORT) [reply continuum/task_set.py:79]: get_rand_samples?

File: continuum/task_set.py (hunk 14 above)
Comment (TLESORT) [reply continuum/task_set.py:141]: seems like a valid point ^^

Comment (TLESORT) [reply continuum/task_set.py:141]: [patch_pr29 7626f16]

File: continuum/scenarios/permutations.py (hunk 11 above)
Comment (TLESORT) [reply continuum/scenarios/permutations.py:34]: I think, it is reproducible now
https://github.com/Continvvm/continuum/blob/master/continuum/scenarios/permutations.py#L32-L34

File: continuum/scenarios/rotations.py (hunk 6 above)
Comment (TLESORT) [reply continuum/scenarios/rotations.py:43]: [patch_pr29 54e3358]

File: README.md (hunk 40 above)
[the remark this answers is not in the document]
Comment (arthurdouillard) [reply README.md:26]: Ok

File: continuum/scenarios/rotations.py (hunk 7 above)
Comment (TLESORT) [reply continuum/scenarios/rotations.py:44]: [patch_pr29 10d9384] and [patch_pr29 f4f5e92]

File: continuum/scenarios/rotations.py (hunk 8 above)
Comment (TLESORT) [reply continuum/scenarios/rotations.py:30]: [patch_pr29 57e50ab]

File: continuum/datasets/imagenet.py (hunk 41 above)
[the remark this answers is not in the document]
Comment (arthurdouillard) [reply continuum/datasets/imagenet.py:62]: I've created an issue (https://github.com/Continvvm/continuum/issues/38).

File: continuum/scenarios/data_incremental.py (hunk 45 above)
[the remark this answers is not in the document]
Comment (arthurdouillard) [reply continuum/scenarios/data_incremental.py:4]: I think it's open to discussion, but we'll see in later updates.

File: continuum/scenarios/transformation_incremental.py (hunk 2 above)
Comment (arthurdouillard) [reply continuum/scenarios/transformation_incremental.py:47]: Yes.

Task 1 is rot10°, task 2 rot20°, but the labels don't change.

File: continuum/scenarios/transformation_incremental.py (hunk 1 above)
Comment (arthurdouillard) [reply continuum/scenarios/transformation_incremental.py:40]: I disagree, the user interface should be simple with no redundance. Actually this redundance in the input arguments is source of bugs from the user code.

To avoid errors, it's on us to do internal checks.

File: continuum/viz.py (hunk 20 above)
Comment (arthurdouillard) [reply continuum/viz.py:7]: I don't doubt it works ;) I'm not sure it is necessary, but nevermind we can keep it that way

File: continuum/scenarios/permutations.py (hunk 12 above)
Comment (TLESORT) [reply continuum/scenarios/permutations.py:56]: You prefer a rule like, "first task has no permutation" rather than "seed 0 is no permutation".
I just want to enable the "no permutation"

File: continuum/scenarios/transformation_incremental.py (hunk 2 above)
Comment (TLESORT) [reply continuum/scenarios/transformation_incremental.py:47]: [patch_pr29 ebee5c8]

File: continuum/scenarios/transformation_incremental.py (hunk 1 above)
Comment (TLESORT) [reply continuum/scenarios/transformation_incremental.py:40]: Ok, it makes sens here modif in beae858
